### PR TITLE
Coinbase maturity part 2

### PR DESCRIPTION
**Pull request by ghost**
Problem found by @beggsdl
------------------------------------------------------------
**beggsdl:** I did this testing by solo mining to my core wallet. Here is a screenshot of the patched v8 wallet confirming a coinbase TX:

image1

**beggsdl:** Here is that same coinbase TX in the 7.17.3 wallet:

image2

**beggsdl:** The transaction detail screen in v8 says it needs to mature for 101 blocks before being spent, but you can see in the Status and Credit, that it is only confirming for 9.

image3

**beggsdl:** This then affects the Immature and Available balances in the v8 wallet (top) compared to the v7 wallet (bottom):

image4

image5

**beggsdl:** I was then able to try to prematurely spend (before 100 confirmations) some of the coinbase TX from the v8 wallet (the v7 wallet did not allow me to do this due to insufficient funds):

image6

**beggsdl:** This 50 DGB transaction is now stuck in mempool of the v8 wallet, even 17 hours after being created.

image7

------------------------------------------------------------
@jaredtate "It looks like there might be a couple of other places where COINBASE_MATURITY_2 needs added properly to fix the issue as @beggsdl mentioned above."

txmempool.cpp
wallet.cpp

https://github.com/DigiByte-Core/digibyte/pull/157#issuecomment-1881677700

------------------------------------------------------------
With this pr. Generated coins must mature 101 blocks before they can be spent.

image9

image10

image11

### CODE REVIEW

**Comment by saltedlolly** [comment]
@beggsdl thanks for identifying the issue and @Jongjan88 thanks for preparing this PR. :)

**Comment by beggsdl** [comment]
Thanks @Jongjan88 for creating this.  I think it is important for this PR to get merged along with PR 157 before RC3.

**Comment by saltedlolly** [comment]
Both your contributions are greatly appreciated!

**Comment by saltedlolly** [comment]
Were there any other places where COINBASE_MATURITY needed fixing?

**Comment by beggsdl** [comment]
> Were there any other places where COINBASE_MATURITY needed fixing?

I have a long list of files related to tests that I feel need to be addressed.  I need to finish going through another set of files to finalize my list.  I can add that as a separate Issue once I am done, hopefully in the next day or so.  What's confusing is that the tests use COINBASE_MATURITY (8) and COINBASE_MATURITY_ORIGINAL (100).

Other than the list I am creating, I think COINBASE_MATURITY  is addressed everywhere.

**Comment by saltedlolly** [comment]
This PR addresses #161

**Review by ycagel — DISMISSED** [review]
cACK. Great work. I would want the blessings of @gto90 and @JaredTate before any merging.

**Review by JaredTate — DISMISSED** [review]
ACK. This PR looks great! Code compiles & client runs. Thanks for doing this @Jongjan88 Awesome work!

**Comment by beggsdl** [comment]
Can someone check this code to make sure it does not need to be addressed also?  Need to make sure this doesn't need to be changed too.
https://github.com/DigiByte-Core/digibyte/blob/e8247b0d596a443eb75f7bd6a9760534970eb25d/src/bench/block_assemble.cpp#L26-L36
image

**Comment by JaredTate** [comment]
> Can someone check this code to make sure it does not need to be addressed also? Need to make sure this doesn't need to be changed too.
> 
> https://github.com/DigiByte-Core/digibyte/blob/e8247b0d596a443eb75f7bd6a9760534970eb25d/src/bench/block_assemble.cpp#L26-L36
> 
> 
> image

We can change the 130 to 200. Great find, although we can have a discussion on what is the appropriate # here. 200 is the default for BTC. This is all in the bench folder, which is for benchmarking. Benchmarking is something I have never used/ done and it is not complete for the entire protocol. DGB codebase has never been even configured for benchmarking in the code. 

This benchmarking code can help us performance test parts of code and have performance data, but it's nothing essential for core protocol and consensus.

More benchmarking info here:https://github.com/DigiByte-Core/digibyte/blob/develop/doc/benchmarking.md

Benchmarking
============

DigiByte Core has an internal benchmarking framework, with benchmarks
for cryptographic algorithms (e.g. SHA1, SHA256, SHA512, RIPEMD160, Poly1305, ChaCha20), rolling bloom filter, coins selection,
thread queue, wallet balance.

Running
---------------------

For benchmarking, you only need to compile `digibyte_bench`.  The bench runner
warns if you configure with `--enable-debug`, but consider if building without
it will impact the benchmark(s) you are interested in by unlatching log printers
and lock analysis.

    make -C src digibyte_bench

After compiling digibyte-core, the benchmarks can be run with:

    src/bench/bench_digibyte

**Comment by beggsdl** [comment]
Thanks for the clarification on that @JaredTate.

File: src/txmempool.cpp [hunk 1]
```diff
@@ -533,10 +533,17 @@ void CTxMemPool::removeForReorg(CChainState& active_chainstate, int flags)
                 const Coin &coin = active_chainstate.CoinsTip().AccessCoin(txin.prevout);
                 if (m_check_ratio != 0) assert(!coin.IsSpent());
                 unsigned int nMemPoolHeight = active_chainstate.m_chain.Tip()->nHeight + 1;
-                if (coin.IsSpent() || (coin.IsCoinBase() && ((signed long)nMemPoolHeight) - coin.nHeight < COINBASE_MATURITY)) {
-                    txToRemove.insert(it);
-                    break;
-                }             
+                    if (coin.nHeight < Params().GetConsensus().multiAlgoDiffChangeTarget) {
```
Comment (gto90) [inline src/txmempool.cpp]: @Jongjan88 , I think we have a formatting error here as the if statement is aligned to the same level as the subsequent if statement.

I think it should be the following:

```
for (const CTxIn& txin : tx.vin) {
    indexed_transaction_set::const_iterator it2 = mapTx.find(txin.prevout.hash);
    if (it2 != mapTx.end())
        continue;
    const Coin &coin = active_chainstate.CoinsTip().AccessCoin(txin.prevout);
    if (m_check_ratio != 0) assert(!coin.IsSpent());
    unsigned int nMemPoolHeight = active_chainstate.m_chain.Tip()->nHeight + 1;
    if (coin.nHeight < Params().GetConsensus().multiAlgoDiffChangeTarget) {
        if (coin.IsSpent() || (coin.IsCoinBase() && ((signed long)nMemPoolHeight) - coin.nHeight < COINBASE_MATURITY)) { 
            txToRemove.insert(it); 
            break; 
        } 
    } else { 
        if (coin.IsSpent() || (coin.IsCoinBase() && ((signed long)nMemPoolHeight) - coin.nHeight < COINBASE_MATURITY_2)) { 
            txToRemove.insert(it); 
            break; 
        }
    }
}

```

**Review by gto90 — CHANGES_REQUESTED** [review]
With one minor formatting fix I think I am prepared to give this a cACK.

File: src/txmempool.cpp (hunk 1 above)
Comment (ghost) [reply src/txmempool.cpp]: Ty. I changed it.

**Review by JaredTate — APPROVED** [review]
ACK. Thank you to everyone for your help on this! Will be great to get this tested on testnet. This code compiles & runs fine for me.

**Review by ycagel — APPROVED** [review]
cACK. Good find @gto90! Thanks for everyone's efforts.

**Review by gto90 — APPROVED** [review]
cACK

Thank you all for your generous contributions!

**Pull request merged**